windows-installer/auto-update.ps1

```diff
@@ -0,0 +1,103 @@
+        Write-EventLog -Source "auto-update.ps1" -LogName "CertbotAutoUpdate" -EventID 1 -EntryType $level -Message $message
+        Write-Host $message
+    }
+
+    function Write-Error($message) {
+        Write-EventLog -Source "auto-update.ps1" -LogName "CertbotAutoUpdate" -EventID 1 -EntryType Error -Message $message
+        throw $message
+    }
+
+    Write-Message "Starting auto-update workflow ..."
+
+    $ErrorActionPreference = 'Stop'
+
+    $installDir = $PSScriptRoot
+    $installerAuthenticodeCertificateThumbprint = "CHANGEME"
+
+    # Get current local certbot version
+    try {
+        $currentVersion = certbot --version
+        $currentVersion = $currentVersion -replace '^certbot (\d+\.\d+\.\d+).*$', '$1'
+        $currentVersion = [System.Version]"$currentVersion"
+    } catch {
+        Write-Message @"
+An error occured while fetching the current local certbot version:
+$_
+Assuming Certbot is not up-to-date.
+"@ "Warning"
+        $currentVersion = "0.0.0"
+    }
+
+    # Get latest remote certbot version
+    try {
+        $result = Invoke-RestMethod -Uri https://api.github.com/repos/certbot/certbot/releases/latest
+        $latestVersion = $result.tag_name -replace '^v(\d+\.\d+\.\d+).*$', '$1'
+        $latestVersion = [System.Version]"$latestVersion"
+    } catch {
+        Write-Error @"
+Could not get the latest remote certbot version. Error was:
+$_
+Aborting auto-upgrade process.
+"@
+    }
+
+    if ($currentVersion -ge $latestVersion) {
+        Write-Message "No upgrade is needed, Certbot is already at the latest version ($currentVersion)."
+    } else {
+        # Search for the Windows installer asset
+        $installerUrl = $null
+        foreach ($asset in $result.assets) {
+            if ($asset.name -match '^certbot-.*installer-win32\.exe$') {
+                $installerUrl = $asset.browser_download_url
+            }
+        }
+
+        if ($null -eq $installerUrl) {
+            Write-Error "Could not find the URL for the latest Certbot for Windows installer."
+        }
+
+        Write-Message "Starting Certbot auto-upgrade from $currentVersion to $latestVersion ..."
+
+        $installerPath = "$env:TMP/certbot-installer-win32.exe"
+        try {
+            # Download the installer
+            Write-Message "Downloading the installer ..."
+            $webClient = New-Object System.Net.WebClient
+            $webClient.DownloadFile($installerUrl, $installerPath)
+
+            # Check installer has a valid signature from the Certbot release team
+            $signature = Get-AuthenticodeSignature $installerPath
+
+            # Uncomment the following lines of code once the Certbot installer is correctly signed.
+    #       if ($signature.Status -ne 'Valid') {
+    #           throw "Downloaded installer has no or invalid Authenticode signature."
+    #       }
+    #       if ($signature.SignerCertificate.Thumbprint -ne $installerAuthenticodeCertificateThumbprint) {
```

Comment:
Do you have a stronger identity check in mind?

Reply:
Like the public key? We will need to always use the same private key, and never loose it over the years though.

Reply:
Yeah I think checking the (hash of the) public key probably makes the most sense. We could do something more complicated like what [the update framework](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md) does and have something like an extra key that we use to sign a file specifying the certs/keys that should be trusted, but I personally think that's probably overkill. certbot-auto only trusts a single key.

I think this design is probably something we should flag for everybody in Mattermost or a weekly meeting in case anyone is strongly in favor of something different, but I doubt they will be.

Reply:
I implemented a check against the Certbot public key used for certbot-auto.